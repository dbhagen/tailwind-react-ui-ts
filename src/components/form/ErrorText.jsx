import React from 'react'
import PropTypes from 'prop-types'

import { Box } from '../primitives'
import { DangerText } from '../typography'

/**
 * @see Renders as [DangerText](#dangertext) component by default
 */
const ErrorText = ({ field: { errorId }, is, ...rest }) => (
  <Box is={is} id={errorId} aria-live="assertive" {...rest} />
)

ErrorText.propTypes = {
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  field: PropTypes.shape({
    errorId: PropTypes.string,
  }),
  size: PropTypes.number,
}

ErrorText.defaultProps = {
  is: DangerText,
  field: {},
  size: 1,
}

export default ErrorText
