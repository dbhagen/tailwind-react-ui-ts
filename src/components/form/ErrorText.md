Component for field level validation error messages that's linked to the input by `aria-describedby` and has `aria-live="assertive"` to announce the error if it's added dynamically.
