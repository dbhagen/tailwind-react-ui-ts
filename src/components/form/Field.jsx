import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Box } from '../primitives'
import { getUniqueID, filterProps } from '../utils'

class Field extends PureComponent {
  constructor(props) {
    const { id } = props

    super(props)

    this.id = id || getUniqueID('field')
  }

  render() {
    const {
      theme,
      is,
      children,
      className,
      hasHelp,
      hasError,
      disabled,
      optionList,
      ...rest
    } = this.props

    const fieldProps = {
      inputId: `${this.id}-input`,
      helpId: hasHelp ? `${this.id}-help` : undefined,
      errorId: hasError ? `${this.id}-error` : undefined,
      labelId: optionList ? `${this.id}-label` : undefined,
      invalid: hasError,
      disabled,
    }

    return (
      <Box
        is={is}
        id={this.id}
        m={{ b: theme.spacing.md }}
        maxW="sm"
        {...filterProps(rest, ['id'])}
      >
        {React.Children.map(children, child =>
          React.cloneElement(child, { field: fieldProps }),
        )}
      </Box>
    )
  }
}

Field.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  className: PropTypes.string,
  hasHelp: PropTypes.bool,
  hasError: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  optionList: PropTypes.bool,
}

Field.defaultProps = {
  is: 'div',
  children: undefined,
  className: undefined,
  hasHelp: false,
  hasError: false,
  disabled: false,
  id: undefined,
  optionList: false,
}

export { Field as component }
export default withTheme(Field)
