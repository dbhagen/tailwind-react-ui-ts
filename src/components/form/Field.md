Wrapping container that passes down a `field` prop object containing ids to be used for accessibility purposes to ensure info text is read out by screen readers
