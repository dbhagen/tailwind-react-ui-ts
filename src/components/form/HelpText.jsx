import React from 'react'
import PropTypes from 'prop-types'

import { Box } from '../primitives'
import { InfoText } from '../typography'

/**
 * @see Renders as [InfoText](#infotext) component by default
 */
const HelpText = ({ is, field: { helpId }, ...rest }) => (
  <Box is={is} id={helpId} {...rest} />
)

HelpText.propTypes = {
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  field: PropTypes.shape({
    helpId: PropTypes.string,
  }),
  size: PropTypes.number,
}

HelpText.defaultProps = {
  is: InfoText,
  field: {},
  size: 1,
}

export default HelpText
