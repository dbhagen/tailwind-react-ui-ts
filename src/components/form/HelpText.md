Wrapper for field help text with id used by input's `aria-describedby` attribute.
