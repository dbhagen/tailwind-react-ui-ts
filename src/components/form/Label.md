Label component that uses the prop object `Field` passes down to apply `htmlFor`.
