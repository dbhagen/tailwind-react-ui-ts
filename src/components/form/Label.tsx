import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Text } from '../primitives'

const Label = ({
  theme,
  is,
  id,
  field: { labelId, inputId, disabled },
  children,
  htmlFor,
  optionList,
  ...rest
}) => (
  <Text
    is={is}
    id={labelId || id}
    inlineBlock
    htmlFor={!optionList ? inputId || htmlFor : undefined}
    m={{ b: theme.spacing.sm }}
    opacity={disabled ? 50 : undefined}
    weight="bold"
    {...rest}
  >
    {children}
  </Text>
)

Label.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  id: PropTypes.string,
  field: PropTypes.shape({
    inputId: PropTypes.string,
    disabled: PropTypes.bool,
  }),
  children: PropTypes.node,
  htmlFor: PropTypes.string,
  optionList: PropTypes.bool,
}

Label.defaultProps = {
  is: 'label',
  id: undefined,
  field: {
    disabled: false,
  },
  children: undefined,
  htmlFor: undefined,
  optionList: false,
}

export { Label as component }
export default withTheme(Label)
