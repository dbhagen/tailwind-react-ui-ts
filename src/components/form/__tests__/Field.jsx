import React from 'react'
import { shallow } from 'enzyme'

import { component as Field } from '../Field'
import { defaultTheme } from '../../theme'

import Label from '../Label'
import TextInput from '../TextInput'
import HelpText from '../HelpText'
import ErrorText from '../ErrorText'

const setup = (testProps = {}) => {
  const props = Object.assign(
    { theme: defaultTheme, hasError: true, hasHelp: true },
    testProps,
  )

  const wrapper = shallow(
    <Field {...props}>
      <Label>Password</Label>
      <HelpText>
        Your password must be 8-20 characters long, contain letters and numbers,
        and must not contain spaces, special characters, or emoji.
      </HelpText>
      <TextInput name="password" type="password" />
      <ErrorText>Please complete</ErrorText>
    </Field>,
  )

  return {
    props,
    wrapper,
  }
}

describe('Field', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
