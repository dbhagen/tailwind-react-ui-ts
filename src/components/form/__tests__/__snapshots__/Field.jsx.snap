// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders matching snapshot 1`] = `
<Box
  id="field-1"
  inline={false}
  inlineBlock={false}
  is="div"
  m={
    Object {
      "b": 4,
    }
  }
  maxW="sm"
>
  <WithTheme(Label)
    field={
      Object {
        "disabled": false,
        "errorId": "field-1-error",
        "helpId": "field-1-help",
        "inputId": "field-1-input",
        "invalid": true,
        "labelId": undefined,
      }
    }
    key=".0"
  >
    Password
  </WithTheme(Label)>
  <HelpText
    field={
      Object {
        "disabled": false,
        "errorId": "field-1-error",
        "helpId": "field-1-help",
        "inputId": "field-1-input",
        "invalid": true,
        "labelId": undefined,
      }
    }
    is={[Function]}
    key=".1"
    size={1}
  >
    Your password must be 8-20 characters long, contain letters and numbers, and must not contain spaces, special characters, or emoji.
  </HelpText>
  <WithTheme(TextInput)
    field={
      Object {
        "disabled": false,
        "errorId": "field-1-error",
        "helpId": "field-1-help",
        "inputId": "field-1-input",
        "invalid": true,
        "labelId": undefined,
      }
    }
    key=".2"
    name="password"
    type="password"
  />
  <ErrorText
    field={
      Object {
        "disabled": false,
        "errorId": "field-1-error",
        "helpId": "field-1-help",
        "inputId": "field-1-input",
        "invalid": true,
        "labelId": undefined,
      }
    }
    is={[Function]}
    key=".3"
    size={1}
  >
    Please complete
  </ErrorText>
</Box>
`;
